File: prototype/cogit/canonical.py
"""Canonical JSON encoding (docs/spec/object-format-v1.md, ADR-0010).

Profile: UTF-8, keys sorted by Unicode code point, minimal escaping,
integers only (|n| <= 2^53 - 1), floats forbidden, no insignificant
whitespace. Python's json module natively matches the escaping and
key-ordering rules, so this wrapper only enforces the value restrictions.
"""

import json

from .errors import UserError

MAX_SAFE_INT = 2**53 - 1


def _validate_value(value, path="$"):
    if isinstance(value, bool) or value is None or isinstance(value, str):
        return
    if isinstance(value, float):
        raise UserError(f"canonical json: float forbidden at {path}")
    if isinstance(value, int):
        if abs(value) > MAX_SAFE_INT:
            raise UserError(f"canonical json: integer out of safe range at {path}")
        return
    if isinstance(value, list):
        for i, item in enumerate(value):
            _validate_value(item, f"{path}[{i}]")
        return
    if isinstance(value, dict):
        for key, item in value.items():
            if not isinstance(key, str):
                raise UserError(f"canonical json: non-string key at {path}")
            _validate_value(item, f"{path}.{key}")
        return
    raise UserError(f"canonical json: unsupported type {type(value).__name__} at {path}")


def canonical_json(value) -> str:
    """Return the canonical JSON text for a JSON-compatible value."""
    _validate_value(value)
    return json.dumps(
        value,
        ensure_ascii=False,
        allow_nan=False,
        sort_keys=True,
        separators=(",", ":"),
    )


def canonical_json_bytes(value) -> bytes:
    return canonical_json(value).encode("utf-8")


def parse_json(text: str):
    """Parse JSON while rejecting floats, so bad values fail at the boundary."""

    def _reject_float(s: str):
        raise UserError(f"canonical json: float forbidden: {s}")

    return json.loads(text, parse_float=_reject_float)

File: crates/cogit-core/src/canonical.rs
//! Canonical JSON profile (ADR-0010, docs/spec/object-format-v1.md):
//! UTF-8, keys sorted by code point, minimal escaping, integers only with
//! |n| <= 2^53 - 1, floats forbidden, no insignificant whitespace.
//!
//! serde_json without `preserve_order` serializes maps via BTreeMap
//! (code-point key order) and escapes exactly like the profile demands
//! (\b \t \n \f \r, lowercase \u00xx for other controls, raw UTF-8 for
//! non-ASCII), so canonicalization is validation + plain `to_string`.

use serde_json::Value;

use crate::error::{CoreError, Result};

pub const MAX_SAFE_INT: u64 = (1 << 53) - 1;

fn validate_value(value: &Value, path: &str) -> Result<()> {
    match value {
        Value::Null | Value::Bool(_) | Value::String(_) => Ok(()),
        Value::Number(n) => {
            if n.is_f64() {
                return Err(CoreError::User(format!(
                    "canonical json: float forbidden at {path}"
                )));
            }
            let in_range = if let Some(i) = n.as_i64() {
                i.unsigned_abs() <= MAX_SAFE_INT
            } else if let Some(u) = n.as_u64() {
                u <= MAX_SAFE_INT
            } else {
                false
            };
            if !in_range {
                return Err(CoreError::User(format!(
                    "canonical json: integer out of safe range at {path}"
                )));
            }
            Ok(())
        }
        Value::Array(items) => {
            for (i, item) in items.iter().enumerate() {
                validate_value(item, &format!("{path}[{i}]"))?;
            }
            Ok(())
        }
        Value::Object(map) => {
            for (key, item) in map {
                validate_value(item, &format!("{path}.{key}"))?;
            }
            Ok(())
        }
    }
}

/// Canonical JSON text for a JSON-compatible value.
pub fn canonical_json(value: &Value) -> Result<String> {
    validate_value(value, "$")?;
    serde_json::to_string(value).map_err(|e| CoreError::User(format!("canonical json: {e}")))
}

pub fn canonical_json_bytes(value: &Value) -> Result<Vec<u8>> {
    Ok(canonical_json(value)?.into_bytes())
}

/// Parse JSON while rejecting floats at the boundary.
pub fn parse_json(text: &str) -> Result<Value> {
    let value: Value = serde_json::from_str(text)
        .map_err(|e| CoreError::User(format!("invalid json: {e}")))?;
    validate_value(&value, "$")?;
    Ok(value)
}
